testsuite/classlibrary/TestCondVar.sc

```diff
@@ -0,0 +1,386 @@
+		this.assertEquals(doneFlags, [false, false], "Done flags before first signalAll");
+
+		c.signalAll;
+		this.momentarilyYieldUntil { counters == [2, 2] };
+		this.assertEquals(doneFlags, [false, false], "Done flags before flags[0] set");
+
+		flags[0] = true;
+		c.signalAll;
+		this.momentarilyYieldUntil { counters == [3, 3] };
+		this.assertEquals(doneFlags, [true, false], "Done flags after flags[0] set");
+
+		flags[1] = true;
+		c.signalAll;
+		this.momentarilyYieldUntil { counters == [3, 4] }; // thread 1 already exited
+		this.assertEquals(doneFlags, [true, true], "Done flags after flags[1] set");
+	}
+
+	// waitFor - preconditions for timeoutBeats
+	test_waitFor_timeoutBeats_badValuesThrowErrors {
+		var c = CondVar();
+
+		this.assertException({ c.waitFor(nil) }, Error, "nil");
+		this.assertException({ c.waitFor(TestCondVarBadIntConversion()) }, Error, "Bad int conversion");
+		this.assertException({ c.waitFor(TestCondVarBadFloatConversion()) }, Error, "Bad float conversion");
+		this.assertException({ c.waitFor(inf) }, Error, "inf");
+		this.assertException({ c.waitFor(0/0) }, Error, "NaN");
+	}
+
+	test_waitFor_timeoutBeats_immediateTimeout {
+		var c = CondVar();
+		// Makes use of the fact that the condition must be evaluated at least twice unless we timeout immediately
+		var checkValue = { |val|
+			var counter = 0;
+			var result = c.waitFor(val) { counter = counter + 1; false };
+			this.assertEquals(counter, 1, "waitFor(%): predicate only evaluated once".format(val));
+			this.assert(result.not, "waitFor(%): result".format(val));
+		};
+
+		checkValue.(0);
+		checkValue.(0.0);
+		checkValue.(-1);
+		checkValue.(-inf);
+
+		// Check good conversions too, these convert to 0 and 0.0 respectively
+		checkValue.(TestCondVarGoodIntConversion());
+		checkValue.(TestCondVarGoodFloatConversion());
+	}
+
+	// waitFor - without predicate
+	test_waitFor_withoutPredicate_signalled {
+		var c = CondVar();
+		fork { c.signalOne };
+		// use assertEquals in case waitFor returns a non-boolean
+		this.assertEquals(c.waitFor(1), true);
+	}
+
+	test_waitFor_withoutPredicate_timedOut {
+		var c = CondVar();
+		// use assertEquals in case waitFor returns a non-boolean
+		this.assertEquals(c.waitFor(this.smallTimeSlice), false);
+	}
+
+	// waitFor - with predicate
+	test_waitFor_withPredicate_immediatelyTrue {
+		var c = CondVar();
+		this.assertEquals(c.waitFor(this.smallTimeSlice) { true }, true);
+	}
+
+	test_waitFor_withPredicate_signaledButFalseSoTimedOut {
+		var c = CondVar();
+		var counter = 0;
+		var waitResult;
+		fork { c.signalOne };
+
+		// This wait has to be a bit longer because we need to ensure there's enough time for the spurious wakeup before
+		// timing out.
+		waitResult = c.waitFor(0.1) {
+			counter = counter + 1;
+			false
+		};
+		this.assertEquals(waitResult, false, "waitResult");
+		// Evaluated 3 times -- when blocked, when signalled, when timed out
+		this.assertEquals(counter, 3, "counter");
```

Comment:
As the comment indicates, this test assumes the CondVar blocked *before* line 297 signaled it. There's no guarantee things will happen in that order. The fork could signal before `CondVar.waitFor` resulting in an incorrect `counter`. On my machine, this test passes. On GA, it appears to fail at least some of the time. We should make this test behave in a predictable manner.

Reply:
I'm resolving this conversation.